velox/functions/sparksql/ArrayRepeat.cpp

```diff
@@ -0,0 +1,184 @@
+ * See the License for the specific language governing permissions and
+ * limitations under the License.
+ */
+#include "velox/expression/DecodedArgs.h"
+#include "velox/expression/VectorFunction.h"
+
+namespace facebook::velox::functions {
+namespace {
+
+class RepeatFunction : public exec::VectorFunction {
+ public:
+  static constexpr int32_t kMaxResultEntries =
+      std::numeric_limits<int32_t>::max() - 15;
+
+  bool isDefaultNullBehavior() const override {
+    // repeat(null, n) returns an array of n nulls.
+    return false;
+  }
+
+  void apply(
+      const SelectivityVector& rows,
+      std::vector<VectorPtr>& args,
+      const TypePtr& outputType,
+      exec::EvalCtx& context,
+      VectorPtr& result) const override {
+    VectorPtr localResult;
+    if (args[1]->isConstantEncoding()) {
+      localResult = applyConstant(rows, args, outputType, context);
+      if (localResult == nullptr) {
+        return;
+      }
+    } else {
+      localResult = applyFlat(rows, args, outputType, context);
+    }
+    context.moveOrCopyResult(localResult, rows, result);
+  }
+
+ private:
+  VectorPtr applyConstant(
+      const SelectivityVector& rows,
+      std::vector<VectorPtr>& args,
+      const TypePtr& outputType,
+      exec::EvalCtx& context) const {
+    const auto numRows = rows.end();
+    auto pool = context.pool();
+
+    auto* constantCount = args[1]->as<ConstantVector<int32_t>>();
+    if (constantCount->isNullAt(0)) {
+      // If count is a null constant, the result should be all nulls.
+      return BaseVector::createNullConstant(outputType, numRows, pool);
+    }
+
+    auto count = constantCount->valueAt(0);
+    // Negative count is like zero count, and empty array will be returned.
+    count = count < 0 ? 0 : count;
+    VELOX_USER_CHECK_LE(
+        count,
+        kMaxResultEntries,
+        "Count argument of array_repeat function must be less than or equal to 2147483632 (MAX_INT32 - 15)");
+
+    const auto totalCount = count * numRows;
+
+    // Allocate new vectors for indices, lengths and offsets.
+    BufferPtr indices = allocateIndices(totalCount, pool);
+    BufferPtr sizes = allocateSizes(numRows, pool);
+    BufferPtr offsets = allocateOffsets(numRows, pool);
+    auto rawIndices = indices->asMutable<vector_size_t>();
+    auto rawSizes = sizes->asMutable<vector_size_t>();
+    auto rawOffsets = offsets->asMutable<vector_size_t>();
+
+    vector_size_t offset = 0;
+    context.applyToSelectedNoThrow(rows, [&](auto row) {
+      rawSizes[row] = count;
+      rawOffsets[row] = offset;
+      std::fill(rawIndices + offset, rawIndices + offset + count, row);
+      offset += count;
+    });
+
+    return std::make_shared<ArrayVector>(
+        pool,
+        outputType,
+        nullptr,
+        numRows,
+        offsets,
+        sizes,
+        BaseVector::wrapInDictionary(nullptr, indices, totalCount, args[0]));
+  }
+
+  VectorPtr applyFlat(
```

Comment:
this name is confusing since vectors may not be flat

Reply:
Just changed the name to applyNonconstant.